hal-x86_64/src/interrupt/mod.rs

```diff
@@ -0,0 +1,419 @@
+            registers: &mut Registers,
+            code: u64,
+        ) {
+            H::double_fault(Context { registers, code });
+        }
+
+        extern "x86-interrupt" fn timer_isr<H: Handlers<crate::X64>>(_regs: &mut Registers) {
+            H::timer_tick();
+            unsafe {
+                PIC.end_interrupt(0x20);
+            }
+        }
+
+        extern "x86-interrupt" fn keyboard_isr<H: Handlers<crate::X64>>(_regs: &mut Registers) {
+            H::keyboard_controller();
+            unsafe {
+                PIC.end_interrupt(0x21);
+            }
+        }
+
+        extern "x86-interrupt" fn test_isr<H: Handlers<crate::X64>>(registers: &mut Registers) {
+            H::test_interrupt(Context {
+                registers,
+                code: (),
+            });
+        }
+        // TODO(eliza): code fault isrs
+
+        self.descriptors[0x20].set_handler(timer_isr::<H> as *const ());
+        self.descriptors[0x21].set_handler(keyboard_isr::<H> as *const ());
+        self.descriptors[69].set_handler(test_isr::<H> as *const ());
+        self.descriptors[14].set_handler(page_fault_isr::<H> as *const ());
+        self.descriptors[8].set_handler(double_fault_isr::<H> as *const ());
+        Ok(())
+    }
+
+    // unsafe fn register_handler_raw<I>(
+    //     &mut self,
+    //     irq: &I,
+    //     handler: *const (),
+    // ) -> Result<(), hal_core::interrupt::RegistrationError>
+    // where
+    //     I: hal_core::interrupt::Interrupt<Ctrl = Self>,
+    // {
+    //     self.descriptors[irq.vector() as usize]
+    //         // .cast_mut::<I::Handler>()
+    //         .set_handler(handler);
+    //     // TODO(eliza): validate this you dipshit!
+    //     Ok(())
+    // }
+}
+
+impl fmt::Debug for Registers {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Registers")
+            .field("instruction_ptr", &self.instruction_ptr)
+            .field("code_segment", &self.code_segment)
+            .field("cpu_flags", &format_args!("{:#b}", self.cpu_flags))
+            .field("stack_ptr", &self.stack_ptr)
+            .field("stack_segment", &self.stack_segment)
+            .finish()
+    }
+}
+
+impl<T> Interrupt<T> {
+    /// Divide-by-zero interrupt (#D0)
+    pub const DIVIDE_BY_ZERO: Interrupt = Interrupt::new_untyped(0);
+
+    pub const DEBUG: Interrupt = Interrupt::new_untyped(1);
+
+    /// Non-maskable interrupt.
+    pub const NMI: Interrupt = Interrupt::new_untyped(2);
+
+    pub const BREAKPOINT: Interrupt = Interrupt::new_untyped(3);
+
+    pub const OVERFLOW: Interrupt = Interrupt::new_untyped(4);
+
+    pub const BOUND_RANGE_EXCEEDED: Interrupt = Interrupt::new_untyped(5);
+
+    pub const INVALID_OPCODE: Interrupt = Interrupt::new_untyped(6);
+
+    /// A device not available exception
+    pub const DEVICE_NOT_AVAILABLE: Interrupt = Interrupt::new_untyped(7);
+
+    // TODO(eliza): can we enforce that this diverges?
```

Comment:
`Interrupt<ErrorCode, Out = !>`?

Reply:
i think this type actually needs to mostly go away